Add function to remove eGPU

ewagner12 opened the pull request:
This adds another function to the script called "remove" which stops the display manager, unloads the driver, removes the PCIe addresses of the eGPU, reloads the driver if necessary, then restarts the display manager. This is purely an enhancement. In my experience, this allows physically unplugging the eGPU anytime after the "remove" process completes. Without this, I would experience a system hang if the eGPU were unplugged. Let me know your experience with unplugging the eGPU, if this works for you or not, or if you have another way.

Note that AMD card users will see the following error message:
`[drm:amdgpu_pci_remove [amdgpu]] *ERROR* Device removal is currently not supported outside of fbcon`
Of the dozens of removals I've tried, I only had one error that required a hard reboot to solve. It seems like [patches](https://lists.freedesktop.org/archives/dri-devel/2020-May/265386.html) are in the works for better AMD removal (that might not even need restarting X). As it stands with the current kernels, all the steps I've added are strictly necessary for unplugging and they should remain the "safe" option even if better unplugging support comes later.

Also speaking of the kernel, I tried with a couple of old mainline kernels and found 5.0+ to be the requirement. Earlier ones did not work, but later worked fine on both the Linux Mint and Ubuntu systems I tested. Again, let me know what you think of this. I'd like to see it tested with more laptop configurations especially dGPU+eGPU. As it is I would consider this to be a "beta" feature that the user can call only if they want.

ewagner12 committed:
Enhance GPU removal

Added code to remove the PCIe bus IDs of the GPU after switching to the internal mode and reloading. Combined with unloading and reloading the driver, this allows both Nvidia and AMD GPUs to be physically unplugged after the display manager reloads. Previously this would result in a system hang. Tested on kernel 5.4.

ewagner12 committed:
Fix boot bug

Only remove GPU PCIe addresses when user requests switch to internal. Automatic boot function should not trigger removal.

ewagner12 committed:
Create separate "remove" function

Made removing the pcie addresses a separate function. Warns the user before executing.

ewagner12 committed:
Formatting

Add new function to help section and added a space

ewagner12 committed:
Add check before driver reloading

Checks if any devices need the GPU drivers (amdgpu or nvidia) before automatically reloading. Reloads the driver only if any devices says it needs it. Prevents errors and unnecessary module loading.

karli-sjoberg commented:
Wow dude, this is awe, and then some!!!

Unfortunately, I'm smack in the middle of moving to another house right now and the P50 has been temporarily assigned as "datacenter" during this time, haha!

I promise you though that I will test this out, merge and push it, it's just going to take a while longer! I'll get back to this as soon as I'm settled in. Now I have even more to look forward to than I had before! Please, do remind me if it takes too long, I'd hate to just let this be left to collect dust in the middle of GitHub!

Happy summer!

karli-sjoberg commented:
I've looked through the code and it looks beautiful, can't wait to test this out!

ewagner12 commented:
Thanks! Good luck with the move.

ewagner12 committed:
Check if drivers are in use

Adds a check for if the drivers are still in use before unloading them. Previously an application like Folding @ Home could cause the reload to stop at a black screen. Now if the drivers are still in use after the display manager stops, then the display manager is simply reloaded and an error message is given without removing the drivers or egpu.

ewagner12 commented:
Hey dude, just giving this PR a bump and seeing if you've had time to test it out.

karli-sjoberg commented:
No, but thank you for doing exactly what I told you, haha!

Turns out the garage has *just* too low ceiling to fit my 42U rack so naturally, I've started the only sane, logical thing any other person would do and have begun to build a insulated storage shack outside the house, as a server room, of sorts- not kidding! I still have this this PR very much in mind and will pick it up as soon as my "data center" is operational :sunglasses:

karli-sjoberg commented:
Just updating on how things are going; rrreeeaaalllyyy slow... I'm starting to think that the best thing to do is either of two things, depending on which could be done fastest:
1) Port "gswitch" to openSUSE, I've switched camps in favor of the green Geeko :smile: Over "here", folks have built in sort of the same functionality in "nvidia-prime", which is just a shell script, but it's not as convenient as "gswitch".
2) Buy or salvage an NVME drive to be able to install Kubuntu.
The P50 is still acting datacenter so in either case, I'll have to test this out on my X280 instead. Kind of a bummer since I wont be able to test the full i/d/e-GPU experience and settle for i/e-GPU integration, but it's better than nothing :shrug:

ewagner12 commented:
No worries, I am also actually in the process of moving now so I get how it is. I think we don't have anyone else on openSUSE so it could be interesting to see gswitch ported to that.

karli-sjoberg commented:
Duuude, the pain! It's at least as painful as packaging for Debian or Ubuntu but man, I've done it! :smile: 

Sad thing is that they're a little different. SUSE only checks for xorg configs under '/etc/X11/xorg.conf.d', so I'm going to have to make the path to the linked config into a variable. But before I do, I will merge this sucker to avoid having to work out the differences later
